test/server/cards/02_SHD/leaders/FinnThisIsARescue.spec.ts

```diff
@@ -0,0 +1,153 @@
+describe('Finn, This is a Rescue', function () {
+    integration(function (contextRef) {
+        describe('Finn\'s undeployed ability', function () {
+            beforeEach(function () {
+                contextRef.setupTest({
+                    phase: 'action',
+                    player1: {
+                        hand: ['entrenched'],
+                        leader: { card: 'finn#this-is-a-rescue', deployed: false },
+                        groundArena: [{ card: 'battlefield-marine', upgrades: ['jedi-lightsaber'] }],
+                        resources: 4
+                    },
+                    player2: {
+                        hand: ['top-target'],
+                        groundArena: ['wampa'],
+                        resources: 5
+                    }
+                });
+            });
+
+            it('should defeat a friendly upgrade and give a shield token', function () {
+                const { context } = contextRef;
+
+                // Scenario 1: Defeat a friendly upgrade on a friendly unit
+                context.player1.clickCard(context.finn);
+                expect(context.player2).toBeActivePlayer();
+                expect(context.battlefieldMarine).toHaveExactUpgradeNames(['shield']);
+                expect(context.jediLightsaber).toBeInZone('discard');
+
+                context.finn.exhausted = false;
+                context.player2.passAction();
+
+                // Scenario 2: Defeat a friendly upgrade on an opponent's unit
+                context.player1.clickCard(context.entrenched);
+                context.player1.clickCard(context.wampa);
+                expect(context.wampa).toHaveExactUpgradeNames(['entrenched']);
+                context.player2.passAction();
+                context.player1.clickCard(context.finn);
+                // There are now two friendly upgrades (entrenched and shield token), so we are prompted to select one
+                context.player1.clickCard(context.entrenched);
+                expect(context.player2).toBeActivePlayer();
+                expect(context.wampa).toHaveExactUpgradeNames(['shield']);
+                expect(context.entrenched).toBeInZone('discard');
+            });
+
+            it('should not defeat an opponent\'s upgrade', function () {
+                const { context } = contextRef;
+
+                context.player1.passAction();
+
+                context.player2.clickCard(context.topTarget);
+                context.player2.clickCard(context.battlefieldMarine);
+                expect(context.battlefieldMarine).toHaveExactUpgradeNames(['jedi-lightsaber', 'top-target']);
```

Comment:
We can join this into the above test by having player2 attach the Top Target at the beginning and confirm that it's never selectable. Same note for the deployed version of this test